Sources/Regex/Atom.swift

```diff
@@ -0,0 +1,474 @@
+
+  /// A built-in escaped character
+  ///
+  /// Literal escapes: \n, \t ...
+  /// Character classes: \s, \w ...
+  /// \n, \s, \Q, \b, \A, \K, ...
+  case escaped(EscapedBuiltin) // TODO: expand this out
+
+  /// A control character
+  ///
+  /// \cx, \C-x, \M-x, \M-\C-x, ...
+  case keyboardControl(Character)
+  case keyboardMeta(Character)        // Oniguruma
+  case keyboardMetaControl(Character) // Oniguruma
+
+  /// A named set (using POSIX syntax)
+  ///
+  /// [:...:], [:^...:]
+  case named(POSIXSet)
+
+  /// .
+  case any
+
+  /// ^
+  case startOfLine
+
+  /// $
+  case endOfLine
+
+  // References
+  //
+  // TODO: Haven't thought through these a ton
+  case backreference(Reference)
+  case subpattern(Reference)
+  case condition(Reference)
+
+  /// Meaningless, used for e.g. non-semantic whitespace
+  ///
+  /// TODO: Does this mean we can't be quantified? we should check sooner...
+  case trivia
+}
+
+extension Atom {
+  struct Storage: Hashable {
+    let kind: Atom
+    let loc: SourceRange
+
+    // TODO: would this be useful to anyone?
+    let fromCustomCharacterClass: Bool
+  }
+}
+
+extension Atom {
+
+  // TODO: We might scrap this and break out a few categories so
+  // we can pull in `^`, `$`, and `.`, but we probably want to
+  // just provide API instead, since that can transcend
+  // taxonomies.
+
+  // Characters, character types, literals, etc., derived from
+  // an escape sequence.
+  public enum EscapedBuiltin: Hashable {
+    // TOOD: better doc comments
+
+    // Literal single characters
+
+    /// \a
+    case alarm
+
+    /// \e
+    case escape
+
+    /// \f
+    case formfeed
+
+    /// \n
+    case newline
+
+    /// \r
+    case carriageReturn
+
+    /// \t
+    case tab
+
+    // Character types
+
+    /// \C
+    case singleDataUnit
+
+    /// \d
+    case decimalDigit
+
+    /// \D
+    case notDecimalDigit
+
+    /// \h
+    case horizontalWhitespace
+
+    /// \H
+    case notHorizontalWhitespace
+
+    /// \N
+    case notNewline
+
+    /// \R
+    case newlineSequence
+
+    /// \s
+    case whitespace
+
+    /// \S
+    case notWhitespace
+
+    /// \v
+    case verticalTab
+
+    /// \V
+    case notVerticalTab
+
+    /// \w
+    case wordCharacter
+
+    /// \W
+    case notWordCharacter
+
+    /// \b (from within a custom character class)
+    case backspace
+
+    // Consumers?
+
+    /// \X
+    case graphemeCluster
+
+    // Assertions
+
+    /// \b (from outside a custom character class)
+    case wordBoundary
+
+    /// \B
+    case notWordBoundary
+
+    // Anchors
+
+    /// \A
+    case startOfSubject
+
+    /// \Z
+    case endOfSubjectBeforeNewline
+
+    /// \z
+    case endOfSubject
+
+    /// \G
+    case firstMatchingPositionInSubject
+
+    // Other
+
+    /// \K
+    case resetStartOfMatch
+
+    // Oniguruma
+
+    /// \O
+    case trueAnychar
+
+    /// \y
+    case textSegment
+
+    /// \Y
+    case notTextSegment
+  }
+}
+
+extension Atom.EscapedBuiltin {
```

Comment:
I'm not convinced we want this nesting vs just flattening this into `Atom` and have a computed property for `isEscaped`. Flatter enums are generally better, and I like having the `atom` case in the AST to contrast with the `indirect` cases, etc.